modyn/selector/internal/selector_strategies/presampling_strategies/label_balanced_presampling_strategy.py

```diff
@@ -0,0 +1,141 @@
+from typing import Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies.presampling_strategies import AbstractPresamplingStrategy
+from sqlalchemy import Select, asc, func, select
+from sqlalchemy.sql.expression import func as sql_func
+
+
+def get_fair_share(capacity: int, requests: list[int]) -> int:
+    """
+    Simple Fair Sharing algorithm.
+    Here it's used to determine how many samples gets each class (min(fair_share, num_samples)
+    """
+    fair_share = int(capacity / len(requests))
+    below_fair_share = [requests.index(el) for el in requests if el <= fair_share]
+
+    if len(below_fair_share) != 0 and len(requests) != len(below_fair_share):
+        new_capacity = capacity - sum(requests[i] for i in below_fair_share)
+        new_requests = [requests[i] for i in range(len(requests)) if i not in below_fair_share]
+        return get_fair_share(new_capacity, new_requests)
+    return fair_share
+
+
+def get_fair_share_predicted_total(fair_share: int, requests: list[int]) -> int:
+    return sum(min(fair_share, req) for req in requests)
+
+
+class LabelBalancedPresamplingStrategy(AbstractPresamplingStrategy):
+    def __init__(self, presampling_config: dict, modyn_config: dict, pipeline_id: int, maximum_keys_in_memory: int):
+        super().__init__(presampling_config, modyn_config, pipeline_id, maximum_keys_in_memory)
+
+        self.force_required_target_size = presampling_config.get("force_required_target_size", False)
```

Comment:
Target size is enforced -- if we want 500 points, then 500 points are returned (even if not perfectly balanced across classes)